### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by dalf**
## What does this PR do?
See commits 

image

Tooltip example for the "Response time" column:
image

Tooltip example for the "Reliability" column:
image

 * Selected language, SafeSearch, Time range columns: a warning is shown when the feature is supported but the checker test fails.
 * Response time shows the median, P80, P95.
 * Reliability column is based on the checker and `searx.metrics.error_recorder`.

---

Related to https://github.com/searxng/searxng/wiki/Milestones#milestone-14---better-statistics-about-the-engines

### CODE REVIEW

**Comment by dalf** [comment]
@return42 can you give a look ?

**Comment by return42** [comment]
> @return42 can you give a look ?

Are you finished? .. or did expect more rebase & force-push in the next time?

**Comment by dalf** [comment]
I think it is okay for a first review.

**Comment by return42** [comment]
> Selected language, SafeSearch, Time range columns: a warning is shown when the feature is supported but the checker test fails.

How did you mock up a test? / I see bing (!) in your screenshot ..

**Comment by dalf** [comment]
Start the checker by sending a SIGUSR1 signal to the Python process:
https://github.com/searxng/searxng/blob/b7848e342273526192a3392dcfd8c291196506ac/searx/search/checker/background.py#L97-L100

Otherwise:

* See to `False`. https://github.com/searxng/searxng/blob/b7848e342273526192a3392dcfd8c291196506ac/searx/settings.yml#L108
* Uncomment: https://github.com/searxng/searxng/blob/b7848e342273526192a3392dcfd8c291196506ac/searx/settings.yml#L116-L118
* Replace `[300, 1800]` by `[10, 30]`.

----

Once the checker has finished, refresh the `/preferences` page.

A solution to quickly have a result from the checker: remove most of the engines from settings.yml...

File: searx/templates/oscar/preferences.html [hunk 1]
```diff
@@ -1,16 +1,92 @@
...
+        <div class="engine-tooltip text-left" role="tooltip" id="{{engine_name}}_time">{{- "" -}}
+            <p>{{ _('Median') }}: {{ stats[engine_name].time }}</p>{{- "" -}}
+            <p>{{ _('P80') }}: {{ stats[engine_name].rate80 }}</p>{{- "" -}}
+            <p>{{ _('P95') }}: {{ stats[engine_name].rate95 }}</p>{{- "" -}}
+        </div>
+    {%- endif -%}
+</td>
+{%- endmacro -%}
+
+{%- macro engine_time(engine_name, css_align_class) -%}
```
Comment (return42) [inline searx/templates/oscar/preferences.html]: Isn't this macro overwriting the one from line 19?

**Review by return42 — CHANGES_REQUESTED** [review]
Seems perfect to me .. just to bikesheeding, some nits I have:

I made a diff of the two relevant macros in the oscar and simple theme (see below) .. does it make sense to implement both macros in the `searx/templates/__common__/` folder?

If you have a look at the diff .. the main difference is in the duplicate of the macro `macro engine_time`.

```diff
@@ -16,26 +16,8 @@
 {%- endif -%}
 {%- endmacro %}
 
-{%- macro engine_time(engine_name, css_align_class) -%}
-<td class="{{ css_align_class }} {{ 'danger' if stats[engine_name]['warn_time'] else '' }}">
-    {%- if stats[engine_name].time != None -%}
-        <span aria-labelledby="{{engine_name}}_time">
-        {%- if stats[engine_name]['warn_time'] -%}
-            {{icon('exclamation-sign')}}
-        {%- endif -%}
-        {{- stats[engine_name].time -}}
-        </span>{{- "" -}}
-        <div class="engine-tooltip text-left" role="tooltip" id="{{engine_name}}_time">{{- "" -}}
-            <p>{{ _('Median') }}: {{ stats[engine_name].time }}</p>{{- "" -}}
-            <p>{{ _('P80') }}: {{ stats[engine_name].rate80 }}</p>{{- "" -}}
-            <p>{{ _('P95') }}: {{ stats[engine_name].rate95 }}</p>{{- "" -}}
-        </div>
-    {%- endif -%}
-</td>
-{%- endmacro -%}
-
-{%- macro engine_time(engine_name, css_align_class) -%}
-<td class="{{ label }}" style="padding: 2px">{{- "" -}}
+{%- macro engine_time(engine_name) -%}
+<td class="{{ label }}" style="padding: 2px; width: 13rem;">{{- "" -}}
     {%- if stats[engine_name].time != None -%}
     <span class="stacked-bar-chart-value">{{- stats[engine_name].time -}}</span>{{- "" -}}
     <span class="stacked-bar-chart" aria-labelledby="{{engine_name}}_chart" aria-hidden="true">{{- "" -}}
@@ -53,28 +35,29 @@
 </td>
 {%- endmacro -%}
 -{%- macro engine_reliability(engine_name, css_align_class) -%}
+{%- macro engine_reliability(engine_name) -%}
 {% set r = reliabilities.get(engine_name, {}).get('reliablity', None) %}
 {% set checker_result = reliabilities.get(engine_name, {}).get('checker', []) %}
 {% set errors = reliabilities.get(engine_name, {}).get('errors', []) %}
 {% if r != None %}
     {% if r <= 50 %}{% set label = 'danger' %}
     {% elif r < 80 %}{% set label = 'warning' %}
-    {% elif r < 90 %}{% set label = 'default' %}
+    {% elif r < 90 %}{% set label = '' %}
     {% else %}{% set label = 'success' %}
     {% endif %}
 {% else %}
     {% set r = '' %}
 {% endif %}
 {% if checker_result or errors %}
-<td class="{{ css_align_class }} {{ label }}">{{- "" -}}
+<td class="{{ label }}">{{- "" -}}
     <span aria-labelledby="{{engine_name}}_reliablity">
-        {%- if reliabilities[engine_name].checker %}{{ icon('exclamation-sign', 'The checker fails on the some tests') }}{% endif %} {{ r -}}
+        {%- if reliabilities[engine_name].checker %}{{ icon('warning', 'The checker fails on the some tests') }}{% endif %} {{ r -}}
     </span>{{- "" -}}
-    <div class="engine-tooltip text-left" role="tooltip" id="{{engine_name}}_reliablity">
+    <div class="engine-tooltip" style="right: 12rem;" role="tooltip" id="{{engine_name}}_reliablity">
         {%- if checker_result -%}
-        <p>{{ _("Failed checker test(s): ") }} {{ ', '.join(checker_result) }}</p>
+        <p>{{ _("The checker fails on this tests: ") }} {{ ', '.join(checker_result) }}</p>
         {%- endif -%}
+        {%- if errors %}<p>{{ _('Errors:') }}</p>{% endif -%}
         {%- for error in errors -%}
         <p>{{ error }} </p>{{- "" -}}
         {%- endfor -%}
```

File: searx/__init__.py [hunk 2]
```diff
@@ -20,7 +20,6 @@
 from os import environ
 from os.path import realpath, dirname, join, abspath, isfile
 
-
```
Comment (return42) [inline searx/__init__.py]: I think this file does not need to be touched.

File: searx/templates/oscar/preferences.html (hunk 1 above)
Comment (dalf) [reply searx/templates/oscar/preferences.html]: Yes, the macro defined line 19 must be removed.

File: searx/static/themes/simple/less/toolkit.less [hunk 3]
```diff
@@ -465,4 +473,62 @@ select {
...
+  flex-wrap: nowrap;
+  align-items: center;
+  display: inline-flex;
+}
+
+.stacked-bar-chart-value {
+  width: 3rem;
+  display: inline-block;
+  position: absolute;
+  padding: 0 0.5rem;
```
Comment (return42) [inline searx/static/themes/simple/less/toolkit.less]: nit: trailing whitespace can be removed

File: searx/templates/oscar/preferences.html (hunk 1 above)
Comment (dalf) [reply searx/templates/oscar/preferences.html]: Fixed baff1cbbab8a72155823c3186971a5f56d7a92f2

File: searx/static/themes/simple/less/toolkit.less (hunk 3 above)
Comment (dalf) [reply searx/static/themes/simple/less/toolkit.less]: Fixed baff1cbbab8a72155823c3186971a5f56d7a92f2

File: searx/__init__.py (hunk 2 above)
Comment (dalf) [reply searx/__init__.py]: Fixed baff1cbbab8a72155823c3186971a5f56d7a92f2

**Review by return42 — APPROVED** [review]

**Comment by return42** [comment]
FYI: merged into https://github.com/return42/searx-next/commit/03f30f3244

**Pull request merged**